sdcm/chaos_mesh.py

```diff
@@ -0,0 +1,62 @@
+import json
+import logging
+
+from sdcm.cluster_k8s import KubernetesCluster
+from sdcm.utils.k8s import HelmValues, KubernetesOps
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ChaosMesh:  # pylint: disable=too-few-public-methods
+    NAMESPACE = "chaos-mesh"
+    VERSION = "2.5.0"
+    HELM_SETTINGS = {'dashboard.create': True, 'dnsServer.create': True}
+
+    def __init__(self):
+        self._k8s_cluster = None
+
+    def initialize(self, k8s_cluster: KubernetesCluster) -> None:
+        """Installs chaos-mesh on k8s cluster and prepares for future k8s chaos testing."""
+        if self._k8s_cluster is not None:
+            LOGGER.warning("Chaos Mesh is already initialized. It should have not been initialized twice.")
+            return
+        self._k8s_cluster = k8s_cluster
+        self._install()
+
+    def _install(self):
+        if self._k8s_cluster.kubectl("kubectl get ns chaos-mesh", ignore_status=True).ok:
```

Comment:
namespace isn't enough, we should wait

Reply:
number of pods for chaos-mesh varies per configuration and number of k8s nodes. It's not easy to predict this number and use this function.
Here I assumed install procedure always succeeds, so verifying existence of chaos-mesh ns would be enough.

Reply:
If it's hard to predict, how do you do that in the end of this function ? I.e. you should have the information, anyhow doesn't the documentation of chaos mesh suggests a different way to wait for it to be ready ?

Also assuming namespace available, mean it's fully installed, is that cause of the atomic variable?

Reply:
At first, we can check that all the pods we have in the namespace are ready and running. There must not be dead pods.
At second, helm chart must support setting number of replicas, so, lets configure it explicitly.
At third. We can read info about replicas number and use it for checking pods readiness.

My choice would be either 2nd or 3rd variant.

But some of them must be done here.

Reply:
In the end, in `_wait_for_chaos_mesh_readiness` I get the number of scheduled pods and wait for that number of pods to be ready. But this approach may fail in case of pods to be scheduled later, after I get number of pods - I was afraid of this scenario and decided to use `atomic` helm argument. And I find this 'atomic' approach a correct one, as we utilize helm function to wait for chart to be ready, instead of a bit hacky `wait_for_pods_readiness` function.

Reply:
> If it's hard to predict, how do you do that in the end of this function ? I.e. you should have the information, anyhow doesn't the documentation of chaos mesh suggests a different way to wait for it to be ready ?
> 
> Also assuming namespace available, mean it's fully installed, is that cause of the atomic variable?

In the end, in _wait_for_chaos_mesh_readiness I get the number of scheduled pods and wait for that number of pods to be ready. But this approach may fail in case of pods to be scheduled later, after I get number of pods - I was afraid of this scenario and decided to use atomic helm argument. And I find this 'atomic' approach a correct one, as we utilize helm function to wait for chart to be ready, instead of a bit hacky wait_for_pods_readiness function.


> At first, we can check that all the pods we have in the namespace are ready and running. There must not be dead pods. At second, helm chart must support setting number of replicas, so, lets configure it explicitly. At third. We can read info about replicas number and use it for checking pods readiness.
> 
> My choice would be either 2nd or 3rd variant.
> 
> But some of them must be done here.

I don't think chaos-mesh's helm chart support replicas - there's one daemon pod for each k8s node and some pods existence depends on configuration (like dashboard). I think current approach of making sure all the pods in chaos-mesh namespace are in running state is enough. Do you know the case when it will not work?

Reply:
Need to specify the affinity rules for the `chaos-mesh` helm chart. 
After it we will know for sure the number of K8S nodes where it must land. And, as a result, we know for sure the number of pods.

Reply:
I think this is handled by chart itself - somehow it installed daemons automatically on all the nodes without setting it.
I'll do more testing, if works as is, maybe we don't need to specify it?

Reply:
It installs on all the nodes. We do not need to have it on all the nodes. We are going to apply it only on Scylla ones.
Actually, it is even easier -> Specify affinity rule for Scylla nodes and use "n_db_nodes" option for getting number of expected pods.
There are a lot of examples for adding affinity rules in the `sdcm/cluster_k8s/__init__.py`.

Reply:
I don't see the benefit here - these pods are very light, installation of this chart with wait for pods to be up took 38 seconds on EKS. Adding affinity logic adds complexity. And still 'n_db_nodes' is not everything, there are chaos-controller-managers and optional dashboard (which I presume will be disabled soon - I keep it just for playing with it)

Reply:
@soyacz adding affinity rules is not `adding complexity`, it is adding `required configuration`.
We `must` do it exactly because we should avoid what you are `ok` with -> creation of lots of redundant pods.
Always think about scale. There always limitations. We should avoid any redundant pod because it is not only resources, but also garbage in logs we gather later.

And, setting of affinity rules is pretty easy and we already have lots of such cases in the K8S code.

Here is what you need:
```
from sdcm.utils.k8s import get_helm_pool_affinity_values

HELM_SETTINGS = get_helm_pool_affinity_values(k8s_kluster.POOL_LABEL_NAME, k8s_kluster.SCYLLA_POOL_NAME)
HELM_SETTINGS |= {
    'dashboard.create': True,
    'dnsServer.create': True,
}
```